pkg/plugins/policies/core/matchers/validators/validator.go

```diff
@@ -34,18 +41,62 @@ func ValidateTargetRef(
 		case common_api.MeshSubset:
 			verr.Add(disallowedField("name", ref.Name, refKind))
 			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
-		case common_api.MeshService:
+			verr.Add(validateTags(ref.Tags))
+		case common_api.MeshService, common_api.MeshHTTPRoute:
 			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(validateName(ref.Name))
 			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
 			verr.Add(disallowedField("tags", ref.Tags, refKind))
-		case common_api.MeshServiceSubset:
+		case common_api.MeshServiceSubset, common_api.MeshGateway:
 			verr.Add(requiredField("name", ref.Name, refKind))
+			verr.Add(validateName(ref.Name))
 			verr.Add(disallowedField("mesh", ref.Mesh, refKind))
+			verr.Add(validateTags(ref.Tags))
 		}
 	}
 	return verr
 }
 
+func validateTags(tags map[string]string) validators.ValidationError {
+	res := validators.ValidationError{}
+	for key, value := range tags {
+		if key == "" {
+			res.Violations = append(res.Violations, validators.Violation{
```

Comment:
Using `res.AddViolation` is a little easier. This enables you to add the key in the error:

```suggestion
			res.AddViolation(validators.RootedAt("tags").Key(key), validators.MustNotBeEmpty))
```

Note we also use existing error messages defined in:

https://github.com/kumahq/kuma/blob/573a9a5f8b569a6d5606bdab384156c34420e022/pkg/core/validators/messages.go#L8-L21

Reply:
addressed